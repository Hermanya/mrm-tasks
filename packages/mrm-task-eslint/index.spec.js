/* eslint-disable no-console */
'use strict';

jest.mock('fs');
jest.mock('mrm-core/src/util/log', () => ({
	added: jest.fn(),
}));
jest.mock('mrm-core/src/npm', () => ({
	install: jest.fn(),
	uninstall: jest.fn(),
}));

const { install } = require('mrm-core');
const { getConfigGetter } = require('mrm');
const vol = require('memfs').vol;
const task = require('./index');

const json = o => JSON.stringify(o, null, '  ');

const packageJson = json({
	name: 'unicorn',
	scripts: {
		test: 'jest',
	},
});

afterEach(() => {
	vol.reset();
	install.mockClear();
});

it('should add ESLint', () => {
	vol.fromJSON({
		'/package.json': packageJson,
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()).toMatchSnapshot();
	expect(install).toBeCalledWith(['eslint']);
});

it('should use a custom preset', () => {
	vol.fromJSON({
		'/package.json': packageJson,
	});

	task(getConfigGetter({ eslintPreset: 'airbnb' }));

	expect(vol.toJSON()['/.eslintrc']).toMatchSnapshot();
	expect(install).toBeCalledWith(['eslint', 'eslint-config-airbnb']);
});

it('should add custom rules', () => {
	vol.fromJSON({
		'/package.json': packageJson,
	});

	task(getConfigGetter({ eslintRules: { 'no-undef': 0 } }));

	expect(vol.toJSON()['/.eslintrc']).toMatchSnapshot();
});

it('should install extra dependencies', () => {
	vol.fromJSON({
		'/package.json': packageJson,
	});

	task(getConfigGetter({ eslintPeerDependencies: ['eslint-plugin-react'] }));

	expect(install).toBeCalledWith(['eslint', 'eslint-plugin-react']);
});

it('should keep custom extensions defined in a package.json script', () => {
	vol.fromJSON({
		'/package.json': json({
			name: 'unicorn',
			scripts: {
				lint: 'eslint --ext .ts',
				test: 'jest',
			},
		}),
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});

it('should remove custom extension if itâ€™s "js" (default value)', () => {
	vol.fromJSON({
		'/package.json': json({
			name: 'unicorn',
			scripts: {
				lint: 'eslint --ext .js',
				test: 'jest',
			},
		}),
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});

it('should add extra rules, parser and extensions for a TypeScript project', () => {
	vol.fromJSON({
		'/package.json': json({
			name: 'unicorn',
			devDependencies: {
				typescript: '*',
			},
		}),
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()).toMatchSnapshot();
	expect(install).toBeCalledWith(['eslint', 'typescript-eslint-parser']);
});
