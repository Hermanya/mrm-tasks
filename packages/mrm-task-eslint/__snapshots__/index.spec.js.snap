// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add ESLint 1`] = `
Object {
  "/.eslintignore": "node_modules/",
  "/.eslintrc": "{
  \\"extends\\": \\"eslint:recommended\\"
}",
  "/.gitignore": ".eslintcache",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {
    \\"pretest\\": \\"npm run lint\\",
    \\"test\\": \\"jest\\",
    \\"lint\\": \\"eslint . --cache --fix\\"
  }
}",
}
`;

exports[`should add custom rules 1`] = `
"{
  \\"extends\\": \\"eslint:recommended\\",
  \\"rules\\": {
    \\"no-undef\\": 0
  }
}"
`;

exports[`should add extra rules, parser and extensions for a TypeScript project 1`] = `
Object {
  "/.eslintignore": "node_modules/",
  "/.eslintrc": "{
  \\"extends\\": \\"eslint:recommended\\",
  \\"parser\\": \\"typescript-eslint-parser\\",
  \\"rules\\": {
    \\"no-undef\\": 0,
    \\"no-unused-vars\\": 0,
    \\"no-useless-constructor\\": 0
  }
}",
  "/.gitignore": ".eslintcache",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"devDependencies\\": {
    \\"typescript\\": \\"*\\"
  },
  \\"scripts\\": {
    \\"lint\\": \\"eslint . --cache --fix --ext .ts,.tsx\\",
    \\"pretest\\": \\"npm run lint\\"
  }
}",
}
`;

exports[`should keep custom extensions defined in a package.json script 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {
    \\"lint\\": \\"eslint . --cache --fix --ext .ts\\",
    \\"pretest\\": \\"npm run lint\\",
    \\"test\\": \\"jest\\"
  }
}"
`;

exports[`should remove custom extension if itâ€™s "js" (default value) 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {
    \\"lint\\": \\"eslint . --cache --fix\\",
    \\"pretest\\": \\"npm run lint\\",
    \\"test\\": \\"jest\\"
  }
}"
`;

exports[`should replace scripts.test.eslint with scripts.lint and scripts.pretest 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {
    \\"pretest\\": \\"npm run lint\\",
    \\"test\\": \\"jest\\",
    \\"lint\\": \\"eslint . --cache --fix --ext .ts\\"
  }
}"
`;

exports[`should use a custom preset 1`] = `
"{
  \\"extends\\": \\"airbnb\\"
}"
`;
